sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +188,309 @@ protected AttestationToken()
+            if (desiredKeyId != null)
+            {
+                // We start looking for candidate signers by looking in the provided attestation signers.
+                // The list of attestation signers trumps
+                foreach (var signer in attestationSigners)
+                {
+                    if (signer.CertificateKeyId == desiredKeyId)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(signer.SigningCertificates.ToArray(), desiredKeyId));
+                        break;
+                    }
+                }
+
+                // We didn't find a certificate from the attestation signers. That means we need to look in the
+                // Header to see if we can find one there.
+                if (candidateCertificates.Count == 0)
+                {
+                    if (Header.JsonWebKey != null)
+                    {
+                        // If the key matches the header's JWK, then the signer is going to be the leaf certificate
+                        // of the certificates in the chain.
+                        if (desiredKeyId == Header.JsonWebKey.Kid)
+                        {
+                            candidateCertificates.Add(new AttestationSigner(
+                                ConvertBase64CertificateArrayToCertificateChain(Header.JsonWebKey.X5C),
+                                desiredKeyId));
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // We don't have a desired Key ID, so we need to try all the possible signing certificates :(.
+
+                // Start with the attestationSigners provided by the caller. Remember that attestation signers
+                // trump all signers in the header.
+                if (attestationSigners != null)
+                {
+                    // Copy the signers to the candidate certificates.
+                    //
+                    // Note: because we don't have a KeyID, we neuter the KeyID in the AttestationSigner.
+                    foreach (var signer in attestationSigners)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(signer.SigningCertificates.ToArray(), null));
+                    }
+                }
+                else
+                {
+                    if (Header.X509CertificateChain != null)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(ConvertBase64CertificateArrayToCertificateChain(Header.X509CertificateChain), null));
+                    }
+                    if (Header.JsonWebKey != null)
+                    {
+                        candidateCertificates.Add(new AttestationSigner(ConvertBase64CertificateArrayToCertificateChain(Header.JsonWebKey.X5C), null));
+                    }
+                }
+            }
+            await Task.Yield();
```

Comment:
I tried that, but when I use Task.FromResult, I get: Since this is an async method, the return expression must be of type 'AttestationSigner[]' rather than 'Task<AttestationSigner[]>'	. And I get the CS1998 about a missing await.